daft/runners/ray_runner.py

```diff
@@ -454,9 +454,14 @@ def next(self, result_uuid: str) -> RayMaterializedResult | StopIteration:
             return StopIteration()
 
         # Common case: get the next result from the thread.
-        result = self.results_by_df[result_uuid].get()
-
-        return result
+        partition_task = self.results_by_df[result_uuid].get()
+        if isinstance(partition_task, SingleOutputPartitionTask):
+            result = partition_task.result_or_wait()
```

Comment:
Note that the consumer of the queue now has to wait on results, because this is no longer a queue of `RayMaterializedResults`, but is now a queue of `SingleOutputPartitionTasks` which may not yet be "done".

This wait is implemented using condition variables, which should be fine I think.